Bluetooth: controller: tx buffer overflow error

**Describe the bug**
I have multiple central devices connecting to a single peripheral asynchronously. The central device writes to one characteristic (around 40 bytes) and reads from another characteristic (around 170 bytes). After few hours of run, the peripheral device runs into the error. The error occurs when the central device tries to read the characteristic.

```
bt_ctlr_hci: Tx Buffer Overflow
bt_hci_core: Data buffer overflow (link type 0x01)
bt_conn: Unable to send to driver (err -55)
```

**To Reproduce**
I've tried developing standalone peripheral and central app to reproduce the issue. But I am unable to do so. I still run into errors on my production code.

**Expected behavior**
the central device should be able to read the characteristic without tx buffer overflow error.

**Impact**
impact on upcoming delivery of the project

**Logs and console output**
cc @vrahane
we've tried to add values of free buffer count into an array and print the array on the error. 
something like this ```tx_mem_free_arr_acq[i_acq].cnt = mem_free_count_get(mem_conn_tx.free);``` in functions ```void *ll_tx_mem_acquire(void)```  and ```void ll_tx_mem_release(void *tx)```

the result is as below:
arr_acp is count in mem_acquire and arr_free is count in mem_release
![Screen Shot 2020-12-01 at 3 31 55 PM](https://user-images.githubusercontent.com/50844088/100831915-91cfad00-341b-11eb-94fa-ad59daa2f694.png)

**Environment (please complete the following information):**
 - OS: MacOS
 - Toolchain: arm-none-eabi-gcc (GNU Tools for Arm Embedded Processors 7-2018-q2-update) 7.3.1 20180622 (release) [ARM/embedded-7-branch revision 261907]
 - Version: Zephyr 2.3.0, b8c78e254ff875680e99c9f131fbe285c4575927

**Additional context**

the above screenshot is with these config options related to tx buffer. 

```
CONFIG_BT_CTLR_TX_BUFFERS=6
CONFIG_BT_CTLR_TX_BUFFER_SIZE=251
CONFIG_BT_L2CAP_TX_BUF_COUNT=6
```

@joerchan The patch above does not work. Can you please provide a git diff instead.

```
/Users/vipul/dev/fw-signal/repos/zephyr/subsys/bluetooth/host/hci_core.c:166:38: error: 'net_buf_num_complete_pool' undeclared (first use in this function)
  166 |  struct net_buf *num_complete_buf = &net_buf_num_complete_pool[0];
      |                                      ^~~~~~~~~~~~~~~~~~~~~~~~~
/Users/vipul/dev/fw-signal/repos/zephyr/subsys/bluetooth/host/hci_core.c:166:38: note: each undeclared identifier is reported only once for each function it appears in
In file included from /Users/vipul/dev/fw-signal/repos/zephyr/include/arch/arm/aarch32/arch.h:27,
                 from /Users/vipul/dev/fw-signal/repos/zephyr/include/arch/cpu.h:19,
                 from /Users/vipul/dev/fw-signal/repos/zephyr/include/kernel_includes.h:38,
                 from /Users/vipul/dev/fw-signal/repos/zephyr/include/kernel.h:17,
                 from /Users/vipul/dev/fw-signal/repos/zephyr/include/zephyr.h:18,
                 from /Users/vipul/dev/fw-signal/repos/zephyr/subsys/bluetooth/host/hci_core.c:10:
/Users/vipul/dev/fw-signal/repos/zephyr/subsys/bluetooth/host/hci_core.c:170:43: error: 'IRQ_ZERO_LATENCY' undeclared (first use in this function)
  170 |  IRQ_DIRECT_CONNECT(MWU_IRQn, 0, mwu_isr, IRQ_ZERO_LATENCY);
      |                                           ^~~~~~~~~~~~~~~~
/Users/vipul/dev/fw-signal/repos/zephyr/include/arch/arm/aarch32/irq.h:117:44: note: in definition of macro 'ARCH_IRQ_DIRECT_CONNECT'
  117 |  z_arm_irq_priority_set(irq_p, priority_p, flags_p); \
      |                                            ^~~~~~~
/Users/vipul/dev/fw-signal/repos/zephyr/subsys/bluetooth/host/hci_core.c:170:2: note: in expansion of macro 'IRQ_DIRECT_CONNECT'
  170 |  IRQ_DIRECT_CONNECT(MWU_IRQn, 0, mwu_isr, IRQ_ZERO_LATENCY);
      |  ^~~~~~~~~~~~~~~~~~
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/local/cmake/cmake --build /Users/vipul/dev/fw-signal/build
make: *** [sdk] Error 1
Vipuls-MacBook-Pro-16-inch-2019:fw-signal vipul$ vim /Users/vipul/dev/fw-signal/repos/zephyr/subsys/bluetooth/host/hci_core.c

repos/zephyr/subsys/bluetooth/host/hci_core.c
166:	struct net_buf *num_complete_buf = &net_buf_num_complete_pool[0];

Press ENTER or type command to continue
```

@joerchan the patch above does not build. Can you please give me the `git diff` instead. I just put it in `hci_core.c`. I also put `nrf.h` in the path.

```
/Users/vipul/dev/fw-signal/repos/zephyr/subsys/bluetooth/host/hci_core.c:167:38: error: 'net_buf_num_complete_pool' undeclared (first use in this function)
  167 |  struct net_buf *num_complete_buf = &net_buf_num_complete_pool[0];
      |                                      ^~~~~~~~~~~~~~~~~~~~~~~~~
/Users/vipul/dev/fw-signal/repos/zephyr/subsys/bluetooth/host/hci_core.c:167:38: note: each undeclared identifier is reported only once for each function it appears in
In file included from /Users/vipul/dev/fw-signal/repos/zephyr/include/arch/arm/aarch32/arch.h:27,
                 from /Users/vipul/dev/fw-signal/repos/zephyr/include/arch/cpu.h:19,
                 from /Users/vipul/dev/fw-signal/repos/zephyr/include/kernel_includes.h:38,
                 from /Users/vipul/dev/fw-signal/repos/zephyr/include/kernel.h:17,
                 from /Users/vipul/dev/fw-signal/repos/zephyr/include/zephyr.h:18,
                 from /Users/vipul/dev/fw-signal/repos/zephyr/subsys/bluetooth/host/hci_core.c:10:
/Users/vipul/dev/fw-signal/repos/zephyr/subsys/bluetooth/host/hci_core.c:171:43: error: 'IRQ_ZERO_LATENCY' undeclared (first use in this function)
  171 |  IRQ_DIRECT_CONNECT(MWU_IRQn, 0, mwu_isr, IRQ_ZERO_LATENCY);
      |                                           ^~~~~~~~~~~~~~~~
/Users/vipul/dev/fw-signal/repos/zephyr/include/arch/arm/aarch32/irq.h:117:44: note: in definition of macro 'ARCH_IRQ_DIRECT_CONNECT'
  117 |  z_arm_irq_priority_set(irq_p, priority_p, flags_p); \
      |                                            ^~~~~~~
/Users/vipul/dev/fw-signal/repos/zephyr/subsys/bluetooth/host/hci_core.c:171:2: note: in expansion of macro 'IRQ_DIRECT_CONNECT'
  171 |  IRQ_DIRECT_CONNECT(MWU_IRQn, 0, mwu_isr, IRQ_ZERO_LATENCY);
      |  ^~~~~~~~~~~~~~~~~~
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/local/cmake/cmake --build /Users/vipul/dev/fw-signal/build
make: *** [sdk] Error 1
```

--- a/subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/cpu.h
+++ b/subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/cpu.h
@@ -16,12 +16,25 @@ static inline void cpu_sleep(void)
 #endif
 }
 
-static inline void cpu_dsb(void)
+static inline void cpu_dmb(void)
 {
-#if defined(CONFIG_CPU_CORTEX_M0) || defined(CONFIG_ARCH_POSIX)
-	/* No need of data synchronization barrier */
-#elif defined(CONFIG_CPU_CORTEX_M4) || defined(CONFIG_CPU_CORTEX_M33)
-	__DSB();
+#if defined(CONFIG_CPU_CORTEX_M)
+	/* NOTE: Refer to ARM Cortex-M Programming Guide to Memory Barrier
+	 *       Instructions, Section 4.1 Normal access in memories
+	 *
+	 *       Implementation: In the Cortex-M processors data transfers are
+	 *       carried out in the programmed order.
+	 *
+	 * Hence, there is no need to use a memory barrier instruction between
+	 * each access. Only a compiler memory clobber is sufficient.
+	 */
+	__asm__ volatile ("" : : : "memory");
+#elif defined(CONFIG_ARCH_POSIX)
+	/* FIXME: Add necessary host machine required Data Memory Barrier
+	 *        instruction alongwith the below defined compiler memory
+	 *        clobber.
+	 */
+	__asm__ volatile ("" : : : "memory");
 #else
 #error "Unsupported CPU."
 #endif

--- a/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c
+++ b/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv.c
@@ -223,7 +223,7 @@ struct pdu_adv *lll_adv_pdu_alloc(struct lll_adv_pdu *pdu, uint8_t *idx)
 		uint8_t first_latest;
 
 		pdu->last = first;
-		cpu_dsb();
+		cpu_dmb();
 		first_latest = pdu->first;
 		if (first_latest != first) {
 			last++;

--- a/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c
+++ b/subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c
@@ -9,15 +9,17 @@
 #include <stddef.h>
 
 #include <toolchain.h>
+#include <soc.h>
 
 #include <sys/util.h>
+
+#include "hal/cpu.h"
+#include "hal/ccm.h"
+#include "hal/radio.h"
 
 #include "util/mem.h"
 #include "util/memq.h"
 #include "util/mfifo.h"
-
-#include "hal/ccm.h"
-#include "hal/radio.h"
 
 #include "pdu.h"
 
@@ -32,7 +34,6 @@
 #define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HCI_DRIVER)
 #define LOG_MODULE_NAME bt_ctlr_lll_conn
 #include "common/log.h"
-#include <soc.h>
 #include "hal/debug.h"
 
 static int init_reset(void);

--- a/subsys/bluetooth/controller/ll_sw/openisa/lll/lll_adv.c
+++ b/subsys/bluetooth/controller/ll_sw/openisa/lll/lll_adv.c
@@ -214,11 +214,11 @@ struct pdu_adv *lll_adv_pdu_alloc(struct lll_adv_pdu *pdu, uint8_t *idx)
 		pdu->last = first;
 		/* FIXME: Ensure that data is synchronized so that an ISR
 		 *        vectored, after pdu->last has been updated, does
-		 *        access the latest value. __DSB() is used in ARM
+		 *        access the latest value. __DMB() is used in ARM
 		 *        Cortex M4 architectures. Use appropriate
 		 *        instructions on other platforms.
 		 *
-		 *        cpu_dsb();
+		 *        cpu_dmb();
 		 */
 		first_latest = pdu->first;
 		if (first_latest != first) {

--- a/subsys/bluetooth/controller/ll_sw/ull_conn.c
+++ b/subsys/bluetooth/controller/ll_sw/ull_conn.c
@@ -6,10 +6,12 @@
 
 #include <stddef.h>
 #include <zephyr.h>
+#include <soc.h>
 #include <device.h>
 #include <bluetooth/bluetooth.h>
 #include <sys/byteorder.h>
 
+#include "hal/cpu.h"
 #include "hal/ecb.h"
 #include "hal/ccm.h"
 #include "hal/ticker.h"
@@ -44,7 +46,6 @@
 #define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HCI_DRIVER)
 #define LOG_MODULE_NAME bt_ctlr_ull_conn
 #include "common/log.h"
-#include <soc.h>
 #include "hal/debug.h"
 
 #if defined(CONFIG_BT_CTLR_USER_EXT)

--- a/subsys/bluetooth/controller/util/memq.c
+++ b/subsys/bluetooth/controller/util/memq.c
@@ -32,8 +32,11 @@
  *   where A[b] means the A'th link-element, whose mem pointer is b.
  */
 
-#include <zephyr/types.h>
 #include <stddef.h>
+
+#include <soc.h>
+
+#include "hal/cpu.h"
 
 #include "memq.h"
 
@@ -97,7 +100,8 @@ memq_link_t *memq_enqueue(memq_link_t *link, void *mem, memq_link_t **tail)
 	/* Update the tail-pointer to point to the new tail element.
 	 * The new tail-element is not expected to point to anything sensible
 	 */
-	*tail = link;
+	cpu_dmb(); /* Ensure data accesses are synchronized */
+	*tail = link; /* Commit: enqueue of memq node */
 
 	return link;
 }

--- a/subsys/bluetooth/controller/util/mfifo.h
+++ b/subsys/bluetooth/controller/util/mfifo.h
@@ -123,6 +123,7 @@ static inline void mfifo_by_idx_enqueue(uint8_t *fifo, uint8_t size, uint8_t idx
 	void **p = (void **)(fifo + (*last) * size); /* buffer preceding idx */
 	*p = mem; /* store the payload which for API 2 is only a void-ptr */
 
+	cpu_dmb(); /* Ensure data accesses are synchronized */
 	*last = idx; /* Commit: Update write index */
 }
 
@@ -189,6 +190,7 @@ static inline uint8_t mfifo_enqueue_get(uint8_t *fifo, uint8_t size, uint8_t cou
  */
 static inline void mfifo_enqueue(uint8_t idx, uint8_t *last)
 {
+	cpu_dmb(); /* Ensure data accesses are synchronized */
 	*last = idx; /* Commit: Update write index */
 }
 

